File: android/src/main/java/com/reactnativewalletmanager/WalletManagerModule.java
package com.reactnativewalletmanager;

import androidx.annotation.NonNull;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.util.Log;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.BaseActivityEventListener; 
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;

import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.pay.PayClient;
import com.google.android.gms.pay.Pay;
import com.google.android.gms.pay.PayApiAvailabilityStatus;

@ReactModule(name = WalletManagerModule.NAME)
public class WalletManagerModule extends ReactContextBaseJavaModule {
    public static final String NAME = "WalletManager";
    private final PayClient walletClient;
    private static final int addToGoogleWalletRequestCode = 1001;
    private Promise addPassPromise;

    public WalletManagerModule(@NonNull ReactApplicationContext reactContext) {
        super(reactContext);
        walletClient = Pay.getClient(reactContext);
        reactContext.addActivityEventListener(mActivityEventListener); 
    }

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    @ReactMethod
    public void canAddPasses(final Promise promise) {
        walletClient.getPayApiAvailabilityStatus(PayClient.RequestType.SAVE_PASSES)
                .addOnSuccessListener(new OnSuccessListener<Integer>() {
                    @Override
                    public void onSuccess(Integer status) {
                        if (status == PayApiAvailabilityStatus.AVAILABLE) {
                            promise.resolve(true);
                        } else {
                            promise.resolve(false);
                        }
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        promise.reject("ERROR", "Error when checking the availability of the Google Wallet API: " + e.getMessage(), e);
                    }
                });
    }

   @ReactMethod
public void addPassToGoogleWallet(String jwt, Promise promise) {
    Activity activity = getCurrentActivity();
    if (activity != null) {
        this.addPassPromise = promise; 
        try {
            walletClient.savePassesJwt(jwt, activity, addToGoogleWalletRequestCode);
            
        } catch (Exception e) {
            e.printStackTrace();
           
            if (e instanceof com.google.android.gms.common.api.ApiException) {
                ApiException apiException = (ApiException) e;
                int statusCode = apiException.getStatusCode();
                String statusMessage = com.google.android.gms.common.api.CommonStatusCodes.getStatusCodeString(statusCode);
                promise.reject("API_ERROR", "Google Pay API error occurred. Status code: " + statusCode + ", Message: " + statusMessage);
            } else {
                promise.reject("GENERAL_ERROR", "Error when saving to Google Wallet: " + e.getMessage(), e);
            }
            this.addPassPromise = null; 
        }
    } else {
        String errorMessage = "Current activity is unavailable.";
        promise.reject("NO_ACTIVITY", errorMessage);
    }
}

    private final ActivityEventListener mActivityEventListener = new BaseActivityEventListener() {
        @Override
        public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
            if (requestCode == addToGoogleWalletRequestCode) {
                if (addPassPromise != null) {
                    switch (resultCode) {
                        case Activity.RESULT_OK:
                            addPassPromise.resolve("Pass successfully added to Google Wallet");
                            break;
                        case Activity.RESULT_CANCELED:
                            addPassPromise.reject("RESULT_CANCELED", "The operation was canceled by the user.");
                            break;
                        default:
                            String error = data != null ? data.getStringExtra("error") : "Unknown error";
                            addPassPromise.reject("UNKNOWN_ERROR", "An error occurred: " + error);
                            break;
                    }
                    addPassPromise = null; 
                }
            }
        }
    };
}

File: android/src/main/java/com/reactnativewalletmanager/WalletManagerModule.kt
package com.reactnativewalletmanager;

import android.app.Activity;
import android.content.Intent;
import com.facebook.react.bridge.BaseActivityEventListener;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.google.android.gms.pay.PayClient;
import com.google.android.gms.pay.Pay;
import com.google.android.gms.pay.PayApiAvailabilityStatus;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.MalformedURLException;

class WalletManagerModule(reactContext: ReactApplicationContext) : NativeWalletManagerSpec(reactContext) {

    private val walletClient: PayClient = Pay.getClient(reactContext)
    private var addPassPromise: Promise? = null
    private val ADD_TO_GOOGLE_WALLET_REQUEST_CODE = 1001

    companion object {
      const val NAME = "NativeWalletManager"
    }

  override fun getName() = NAME

  private val mActivityEventListener = object : BaseActivityEventListener() {
    override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
      if (requestCode == ADD_TO_GOOGLE_WALLET_REQUEST_CODE) {
        addPassPromise?.let { promise ->
          when (resultCode) {
            Activity.RESULT_OK -> promise.resolve("Pass successfully added to Google Wallet")
            Activity.RESULT_CANCELED -> promise.reject("RESULT_CANCELED", "The operation was canceled by the user.")
            else -> {
              val error = data?.getStringExtra("error") ?: "Unknown error"
              promise.reject("UNKNOWN_ERROR", "An error occurred: $error")
            }
          }
          addPassPromise = null
        }
      }
    }
  }

    init {
        reactContext.addActivityEventListener(mActivityEventListener)
    }

    override fun canAddPasses(promise: Promise) {
        walletClient.getPayApiAvailabilityStatus(PayClient.RequestType.SAVE_PASSES)
            .addOnSuccessListener { status ->
                promise.resolve(status == PayApiAvailabilityStatus.AVAILABLE)
            }
            .addOnFailureListener { e ->
                promise.reject(
                    "ERROR",
                    "Error when checking the availability of the Google Wallet API: ${e.message}",
                    e
                )
            }
    }

    override fun addPassToGoogleWallet(jwt: String, promise: Promise) {
        val activity = currentActivity
        if (activity != null) {
            addPassPromise = promise
            try {
                walletClient.savePassesJwt(jwt, activity, ADD_TO_GOOGLE_WALLET_REQUEST_CODE)
            } catch (e: Exception) {
                e.printStackTrace()
                when (e) {
                    is com.google.android.gms.common.api.ApiException -> {
                        val statusCode = e.statusCode
                        val statusMessage = com.google.android.gms.common.api.CommonStatusCodes.getStatusCodeString(statusCode)
                        promise.reject(
                            "API_ERROR",
                            "Google Pay API error occurred. Status code: $statusCode, Message: $statusMessage"
                        )
                    }
                    else -> {
                        promise.reject(
                            "GENERAL_ERROR",
                            "Error when saving to Google Wallet: ${e.message}",
                            e
                        )
                    }
                }
                addPassPromise = null
            }
        } else {
            promise.reject("NO_ACTIVITY", "Current activity is unavailable.")
        }
    }

  override fun showAddPassControllerFromFile(url: String, promise: Promise) {
    promise.reject("UNSUPPORTED_PLATFORM", "This function is not supported on Android.")
  }

  override fun addPassFromUrl(url: String, headers: ReadableMap?, promise: Promise) {
    val activity = currentActivity
    if (activity == null) {
      promise.reject("NO_ACTIVITY", "Current activity is unavailable")
      return
    }

    // Validate URL before starting network request
    val parsedUrl: URL
    try {
      parsedUrl = URL(url)
    } catch (e: MalformedURLException) {
      promise.reject("INVALID_URL", "The URL is invalid")
      return
    }

    // Run network request on background thread
    Thread {
      var connection: HttpURLConnection? = null
      try {
        connection = parsedUrl.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"

        // Add custom headers
        headers?.let { map ->
          val iterator = map.keySetIterator()
          while (iterator.hasNextKey()) {
            val key = iterator.nextKey()
            connection.setRequestProperty(key, map.getString(key))
          }
        }

        val responseCode = connection.responseCode
        if (responseCode !in 200..299) {
          promise.reject("HTTP_ERROR", "HTTP $responseCode")
          return@Thread
        }

        val jwt = connection.inputStream.bufferedReader().use { it.readText() }

        // Save to Google Wallet on main thread
        activity.runOnUiThread {
          addPassPromise = promise
          try {
            walletClient.savePassesJwt(jwt, activity, ADD_TO_GOOGLE_WALLET_REQUEST_CODE)
          } catch (e: Exception) {
            addPassPromise = null
            when (e) {
              is com.google.android.gms.common.api.ApiException -> {
                val statusCode = e.statusCode
                val statusMessage = com.google.android.gms.common.api.CommonStatusCodes.getStatusCodeString(statusCode)
                promise.reject("API_ERROR", "Google Pay API error: $statusCode - $statusMessage")
              }
              else -> {
                promise.reject("GENERAL_ERROR", e.message ?: "Failed to save pass")
              }
            }
          }
        }
      } catch (e: Exception) {
        promise.reject("NETWORK_ERROR", e.message ?: "Network request failed")
      } finally {
        connection?.disconnect()
      }
    }.start()
  }

  override fun hasPass(cardIdentifier: String, serialNumber: String?, promise: Promise) {
    promise.reject("UNSUPPORTED_PLATFORM", "This function is not supported on Android.")
  }

  override fun removePass(cardIdentifier: String, serialNumber: String?, promise: Promise) {
    promise.reject("UNSUPPORTED_PLATFORM", "This function is not supported on Android.")
  }

  override fun viewInWallet(cardIdentifier: String, serialNumber: String?, promise: Promise) {
    promise.reject("UNSUPPORTED_PLATFORM", "This function is not supported on Android.")
  }
}
